src/kubernetes_cluster/proof/kubernetes_api_liveness.rs

```diff
@@ -223,6 +246,209 @@ pub open spec fn no_req_before_rest_id_is_in_flight(rest_id: RestId) -> StatePre
+/// created after some points. Here, "requirements" takes two parameters, the new message and the prime state. In many cases,
+/// It's only related to the message.
+/// 
+/// In detail, we have to show two things:
+///     a. Newly created api request message satisfies requirements: s.in_flight(msg) /\ s_prime.in_flight(msg) ==> requirements(msg, s_prime).
+///     b. The requirements, once satisfied, won't be violated as long as the message is still in flight:
+///         s.in_flight(msg) /\ requirements(msg, s) /\ s_prime.in_flight(msg) ==> requirements(msg, s_prime).
+/// 
+/// Previously, when "requirements" was irrelavant to the state, b will be sure to hold. Later, we find that "requirements" in some 
+/// case does need some information in the state. So we add state as another parameter and requires the caller of the lemma
+/// lemma_true_leads_to_always_every_in_flight_req_msg_satisfies to prove b. is always satisfied. In order not to make those cases
+/// where "requirements" has nothing to do with state more difficult, we combine a. and b. together.
+/// 
+/// Therefore, we have the following predicate. As is easy to see, this is similar as:
+///     (s.in_flight(msg) ==> requirements(msg, s)) ==> (s_prime.in_flight(msg) ==> requirements(msg, s_prime))
+/// If we think of s.in_flight(msg) ==> requirements(msg, s) as an invariant, it is the same as the proof of invariants in previous
+/// proof strategy.
+pub open spec fn every_new_req_msg_if_in_flight_then_always_satisfies(requirements: FnSpec(Message, Self) -> bool) -> ActionPred<Self> {
+    |s: Self, s_prime: Self| {
+        forall |msg: Message|
+            (!s.message_in_flight(msg) || requirements(msg, s))
+            && #[trigger] s_prime.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_APIRequest()
+            ==> requirements(msg, s_prime)
+    }
+}
+
+pub open spec fn every_in_flight_req_msg_satisfies(requirements: FnSpec(Message, Self) -> bool) -> StatePred<Self> {
+    |s: Self| {
+        forall |msg: Message|
+            #[trigger] s.message_in_flight(msg)
+            && msg.dst.is_KubernetesAPI()
+            && msg.content.is_APIRequest()
+            ==> requirements(msg, s)
+    }
+}
+
+/// This lemma shows that if spec ensures every newly created Kubernetes api request message satisfies some requirements,
+/// the system will eventually reaches a state where all Kubernetes api request messages satisfy those requirements.
+///
+/// To require "every newly create Kubernetes api request message satisfies some requirements", we use a FnSpec (i.e., a closure)
+/// as parameter which can be defined by callers and require spec |= [](every_new_req_msg_if_in_flight_then_always_satisfies(requirements)).
+///
+/// The last parameter must be equivalent to every_in_flight_req_msg_satisfies(requirements)
+pub proof fn lemma_true_leads_to_always_every_in_flight_req_msg_satisfies(
```

Comment:
Actually, I would recommend keeping this GREAT lemma even simpler by removing the last parameter from this lemma and just use `Self::every_in_flight_req_msg_satisfies(requirements)` in the postcondition. If the caller needs to convert it to a msg_state_pred, the caller should call `temp_pred_equality`. This lemma should just focus on proving `true ~> []every_in_flight_req_msg_satisfies(requirements)`.